src/toxcore/onion/onion_announce.rs

```diff
@@ -0,0 +1,790 @@
+            self.entries.get(i)
+        } else if self.entries.len() < ONION_ANNOUNCE_MAX_ENTRIES {
+            // adding new entry does not exceed the limit - just add it
+            let pk = entry.pk;
+            self.entries.push(entry);
+            self.entries.sort_by(|e1, e2| dht_pk.distance(&e1.pk, &e2.pk));
+            self.entries.iter().find(|e| e.pk == pk)
+        } else if dht_pk.distance(&entry.pk, &self.entries.last().unwrap().pk) == Ordering::Less {
+            // the farthest entry is farther than new entry - replace it
+            // unwrap can not fail due to previous branch
+            self.entries.last_mut().unwrap().clone_from(&entry);
+            self.entries.sort_by(|e1, e2| dht_pk.distance(&e1.pk, &e2.pk));
+            self.entries.iter().find(|e| e.pk == entry.pk)
+        } else {
+            None
+        }
+    }
+
+    /** Handle `AnnounceRequest` packet and return `AnnounceResponse`.
+
+    If `AnnounceRequest` packet contains valid onion ping id it's considered as
+    announce request. Otherwise it's considered as search request. In case of
+    announce request we try to add new entry to announce list and if succeed
+    return status `AnnounceStatus::Announced`. In case of search request we try
+    to find entry by `search_pk` key from request and if succeed return status
+    `AnnounceStatus::Found`. If announce or search failed we return status
+    `AnnounceStatus::Failed`.
+
+    If request is a search request and we found requested node then
+    `ping_id_or_pk` field in response packet will contain `PublicKey` that
+    should be used to send data packets to requested node. Otherwise it will
+    contain valid onion ping id that should be used to send announce requests
+    to this node.
+
+    Also response packet will contain up to 4 closest to `search_pk` nodes from
+    kbucket. They are used to search closest to long term `PublicKey` nodes to
+    announce.
+
+    */
+    pub fn handle_announce_request(&mut self, request: AnnounceRequest, dht_sk: &SecretKey, kbucket: &Kbucket, addr: SocketAddr) -> Result<AnnounceResponse, Error> {
+        let shared_secret = precompute(&request.inner.pk, dht_sk);
+        let payload = request.inner.get_payload(&shared_secret)?;
+
+        let time = SystemTime::now();
+        let ping_id_1 = self.ping_id(
+            time,
+            request.inner.pk,
+            addr.ip(),
+            addr.port()
+        );
+        let ping_id_2 = self.ping_id(
+            time + Duration::from_secs(PING_ID_TIMEOUT),
+            request.inner.pk,
+            addr.ip(),
+            addr.port()
+        );
+
+        let entry_opt = if payload.ping_id == ping_id_1 || payload.ping_id == ping_id_2 {
+            let entry = OnionAnnounceEntry::new(request.inner.pk, addr.ip(), addr.port(), request.onion_return, payload.data_pk);
+            self.add_to_entries(entry)
+        } else {
+            self.find_in_entries(payload.search_pk)
```

Comment:
Do we need `find_in_entries`, since we try to actually find the entry by pk in `add_to_entries`?

Reply:
Yes, because it's different request - not announce, but search request. In this case we only search node by search_pk.

Reply:
OK